File: NetDeck/CardImageViewPopover.m
//
//  CardImageViewPopover.m
//  Net Deck
//
//  Created by Gereon Steffens on 28.12.13.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "CardImageViewPopover.h"
#import "CardDetailView.h"

#define POPOVER_MARGIN  40 // 20px status bar + 10px top + 10px bottom
#define IPAD_SCREEN_HEIGHT   768

@interface CardImageViewPopover ()

@property Card* card;
@property BOOL showAlt;

@end

static CardImageViewPopover* popover;

static BOOL keyboardVisible = NO;
static CGFloat popoverScale = 1.0;

@implementation CardImageViewPopover

#pragma mark keyboard monitor

+(void)monitorKeyboard
{
    NSNotificationCenter* nc = [NSNotificationCenter defaultCenter];
    
    [nc addObserver:self selector:@selector(showKeyboard:) name:UIKeyboardDidShowNotification object:nil];
    [nc addObserver:self selector:@selector(hideKeyboard:) name:UIKeyboardWillHideNotification object:nil];
}

+(void) showKeyboard:(NSNotification*)notification
{
    keyboardVisible = YES;
    NSValue* value = notification.userInfo [UIKeyboardFrameEndUserInfoKey];
    CGFloat keyboardHeight = value.CGRectValue.size.height;
    popoverScale = (IPAD_SCREEN_HEIGHT - keyboardHeight - POPOVER_MARGIN) / ImageCache.IMAGE_HEIGHT;
    popoverScale = MIN(1.0, popoverScale);
}

+(void) hideKeyboard:(NSNotification*)sender
{
    keyboardVisible = NO;
    popoverScale = 1.0;
    
    if (popover)
    {
        popover.view.transform = CGAffineTransformIdentity;
        popover.preferredContentSize = CGSizeMake(ImageCache.IMAGE_WIDTH, ImageCache.IMAGE_HEIGHT);
    }
}

#pragma mark show/dismiss

+(void)showForCard:(Card *)card fromRect:(CGRect)rect inViewController:(UIViewController *)vc subView:(UIView *)view
{
    if (card == nil || vc == nil)
    {
        return;
    }
    
    NSAssert(popover == nil, @"previous popover still visible?");
    popover = [[CardImageViewPopover alloc] initWithCard:card];
    
    popover.modalPresentationStyle = UIModalPresentationPopover;
    popover.popoverPresentationController.sourceRect = rect;
    popover.popoverPresentationController.sourceView = view;
    popover.popoverPresentationController.permittedArrowDirections = UIPopoverArrowDirectionLeft|UIPopoverArrowDirectionRight;
    popover.popoverPresentationController.delegate = popover;
    popover.preferredContentSize = CGSizeMake((int)(ImageCache.IMAGE_WIDTH*popoverScale), (int)(ImageCache.IMAGE_HEIGHT*popoverScale));
    
    [vc presentViewController:popover animated:NO completion:nil];
}

+(BOOL) dismiss
{
    if (popover)
    {
        [popover dismissViewControllerAnimated:NO completion:nil];
        popover = nil;
        return YES;
    }
    return NO;
}

-(void) popoverPresentationControllerDidDismissPopover:(UIPopoverPresentationController *)popoverPresentationController
{
    popover = nil;
}

- (id)initWithCard:(Card*)card
{
    self = [super initWithNibName:@"CardImageView" bundle:nil];
    if (self)
    {
        self.card = card;
        self.detailView.hidden = YES;
        
        if (keyboardVisible)
        {
            self.view.transform = CGAffineTransformMakeScale(popoverScale, popoverScale);
        }
    }
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.imageView.userInteractionEnabled = YES;
    UITapGestureRecognizer* imgTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(imgTap:)];
    imgTap.numberOfTapsRequired = 1;
    [self.imageView addGestureRecognizer:imgTap];
    
    [self loadCardImage];
}

-(void) imgTap:(UITapGestureRecognizer*)sender
{
    if (UIGestureRecognizerStateEnded == sender.state)
    {
        [CardImageViewPopover dismiss];
    }
}

-(void) loadCardImage
{
    [self.activityIndicator startAnimating];
    [self loadCardImage:self.card];
}

-(void) loadCardImage:(Card*)card
{
    if (card == nil) {
        return;
    }
    [[ImageCache sharedInstance] getImageFor:card
                                     completion:^(Card* card, UIImage* image, BOOL placeholder) {
                                         if ([card.code isEqualToString:self.card.code])
                                         {
                                             [self.activityIndicator stopAnimating];
                                             self.imageView.image = image;
                                             
                                             self.detailView.hidden = !placeholder;
                                             if (placeholder)
                                             {
                                                 [CardDetailView setupDetailViewFromPopover:self card:self.card];
                                             }
                                         }
                                         else
                                         {
                                             [self loadCardImage:self.card];
                                         }
                                     }];

}

@end

File: NetDeck/CardImageViewPopover.swift
//
//  CardImageViewPopover.swift
//  Net Deck
//
//  Created by Gereon Steffens on 28.12.13.
//  Copyright © 2019 Gereon Steffens. All rights reserved.
//

import UIKit
import SwiftyUserDefaults

class CardImageViewPopover: UIViewController, UIPopoverPresentationControllerDelegate, CardDetailDisplay {
    
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var activityIndicator: UIActivityIndicatorView!
    
    @IBOutlet weak var detailView: UIView!
    @IBOutlet weak var cardName: UILabel!
    @IBOutlet weak var cardType: UILabel!
    @IBOutlet weak var cardText: UITextView!
    
    @IBOutlet weak var label1: UILabel!
    @IBOutlet weak var label2: UILabel!
    @IBOutlet weak var label3: UILabel!
    @IBOutlet weak var icon1: UIImageView!
    @IBOutlet weak var icon2: UIImageView!
    @IBOutlet weak var icon3: UIImageView!
    @IBOutlet weak var packLabel: InsetLabel!
    
    @IBOutlet weak var mwlLabel: InsetLabel!
    @IBOutlet weak var mwlRightDistance: NSLayoutConstraint!
    @IBOutlet weak var mwlBottomDistance: NSLayoutConstraint!
    
    private var card: Card!
    private var mwl: MWL!
    
    private static var keyboardMonitor: KeyboardMonitor!
    private static var keyboardObserver: KeyboardObserver!
    
    private static var popover: CardImageViewPopover?
    
    private static var keyboardVisible = false
    private static var popoverScale: CGFloat = 1.0
    
    private static let popoverMargin: CGFloat = 40
    
    // MARK: - keyboard monitor
    
    static func monitorKeyboard() {
        keyboardMonitor = KeyboardMonitor()
        keyboardObserver = KeyboardObserver(handler: keyboardMonitor)
    }
    
    static func showKeyboard(_ info: KeyboardInfo) {
        keyboardVisible = true

        let screenHeight = UIScreen.main.bounds.size.height
        let kbHeight = screenHeight - info.endFrame.origin.y
        popoverScale = (screenHeight - kbHeight - popoverMargin) / CGFloat(ImageCache.height)
        popoverScale = min(1.0, popoverScale)
    }
    
    static func hideKeyboard() {
        keyboardVisible = false
        popoverScale = 1.0
        
        if let p = popover {
            p.view.transform = .identity
            p.preferredContentSize = CGSize(width: ImageCache.width, height: ImageCache.height)
        }
    }
    
    // MARK: - show/dismis
    
    static func show(for card: Card, from rect: CGRect, in vc: UIViewController, subView view: UIView) {
        show(for: card, mwl: nil, from: rect, in: vc, subView: view)
    }
    
    static func show(for card: Card, mwl: MWL?, from rect: CGRect, in vc: UIViewController, subView view: UIView) {
        assert(CardImageViewPopover.popover == nil, "previous popover still visible?")
        
        let popover = CardImageViewPopover()
        popover.card = card
        popover.mwl = mwl ?? Defaults[.defaultMWL]
        
        popover.modalPresentationStyle = .popover
        if let pres = popover.popoverPresentationController {
            pres.sourceRect = rect
            pres.sourceView = view
            pres.permittedArrowDirections = [.left, .right]
            pres.delegate = popover
        }
        popover.preferredContentSize = CGSize(width: Int(ImageCache.width*popoverScale), height: Int(ImageCache.height*popoverScale))
        CardImageViewPopover.popover = popover
        
        vc.present(popover, animated:false, completion:nil)
    }
    
    @discardableResult
    static func dismiss() -> Bool {
        if let p = popover {
            p.dismiss(animated:false, completion:nil)
            popover = nil
            return true
        }
        return false
    }
    
    func popoverPresentationControllerDidDismissPopover(_ popoverPresentationController: UIPopoverPresentationController) {
        CardImageViewPopover.popover = nil
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.detailView.isHidden = true
        if CardImageViewPopover.keyboardVisible {
            self.view.transform = CGAffineTransform(scaleX: CardImageViewPopover.popoverScale, y: CardImageViewPopover.popoverScale)
        }
        
        self.imageView.isUserInteractionEnabled = true
        let imgTap = UITapGestureRecognizer(target: self, action: #selector(self.imgTap(_:)))
        imgTap.numberOfTapsRequired = 1
        self.imageView.addGestureRecognizer(imgTap)
        
        let penalty = card.mwlPenalty(self.mwl)
        
        self.mwlLabel.isHidden = penalty == 0
        self.mwlLabel.text = (self.mwl.universalInfluence ? "+" : "-") + "\(penalty)"
        switch card.type {
        case .event, .hardware, .resource, .program, .ice:
            self.mwlRightDistance.constant = 6
            self.mwlBottomDistance.constant = 124
        case .agenda, .asset, .upgrade, .operation:
            self.mwlRightDistance.constant = 154
            self.mwlBottomDistance.constant = 15
        default:
            self.mwlLabel.isHidden = true
        }
        
        self.packLabel.layer.cornerRadius = 3
        self.packLabel.layer.masksToBounds = true
        self.packLabel.text = nil
        
        self.mwlLabel.layer.cornerRadius = 3
        self.mwlLabel.layer.masksToBounds = true
        self.mwlLabel.font = UIFont.monospacedDigitSystemFont(ofSize: 9, weight: UIFont.Weight.bold)
        
        self.activityIndicator.startAnimating()
        self.loadCardImage(self.card)
    }
    
    @objc func imgTap(_ sender: UITapGestureRecognizer) {
        if sender.state == .ended {
            CardImageViewPopover.dismiss()
        }
    }
    
    private func loadCardImage(_ card: Card) {
        ImageCache.sharedInstance.getImage(for: card) { (card, img, placeholder) in
            if self.card.code == card.code {
                self.activityIndicator.stopAnimating()
                self.imageView.image = img
                self.packLabel.text = card.packName
                self.packLabel.layer.cornerRadius = 3
                self.packLabel.layer.masksToBounds = true
                
                self.detailView.isHidden = true
                if placeholder {
                    CardDetailView.setup(from: self, card: self.card)
                }
            } else {
                self.loadCardImage(self.card)
            }
        }
    }
}

private class KeyboardMonitor: KeyboardHandling {
    func keyboardWillShow(_ info: KeyboardInfo) {
        CardImageViewPopover.showKeyboard(info)
    }
    
    func keyboardWillHide(_ info: KeyboardInfo) {
        CardImageViewPopover.hideKeyboard()
    }
}
